### ISSUE TITLE
Mount with --config read from named pipe

### ISSUE THREAD

**Issue by schlomie** [netheril96/securefs#57]
Would it be possible to mount a folder with `--config` read from a named pipe?

Example, I would like to keep my configuration file kept in an encrypted GPG file that I can write into the named pipe at mount time.

From one terminal:
`
$ securefs m -b --config config.pipe priv pub
`

From a second terminal:
`
$ gpg -d config.asc > config.pipe
`

Doing something similar works with EncFS and gocryptfs, but I get a POSIXException Illegal seek when attempting with securefs.

**Comment by schlomie** [comment]
Hmm.  Digging in to the code, it appears this may not be possible.  The exception appears to be thrown in the implementation of UnixFileStream::size(), invoking std::lseek on the file pointer.  Seekable named pipes are a no-go.

Looks like this type of feature would require some major rewriting...

**Comment by netheril96** [comment]
It doesn't require a major rewrite. I can just handle config files in a special way.

**Comment by antofthy** [comment]
This same problem popped up in EncFS.  It was allowed, but then a update switched from using "boost" to using TinyXML2 for parsing the XML!  But TinyXML2 wanted to first know how big the file was before reading it so it could allocate memory for the file before hand, and did so using "seek()".  Named pipe cannot use seek() as it is a pipeline not a random access file.

That was a BIG pain as I do not store config files in the encrypted directory, but fed that file to EncFS as it started up via a named pipe from a database.  After EncFS has started the named pipe is removed.  This allowed the config to be passed to EncFS securely.  A temporary workaround was to write the config into a tmp file then remove it later, which was much less secure.

EncFS fixed that problem, by reading the file into memory itself using a expanding memory allocation technique. The discussion of this is in the EncFS Issues #253

I strongly suggest you do the same. The only requirement of the XML parser is to read the config file ONCE only. You can get around it by reading it into memory, then parse, (deleting from memory after parsing).  This is is how EncFS does it now.   TinyXML2 also now does similar.

**Comment by netheril96** [comment]
Commit edd62bb51b5f5bc8e71d49c3bfaf57aad75aae6b now adds support for non-seekable files as config. I need more testing on Windows before merging it with master.

**Comment by schlomie** [comment]
I do not have a windows machine from which I can help --- but I did check out the sequential branch, built and was successful in mounting a folder reading the config from a pipe.

This is fantastic!

**Comment by netheril96** [comment]
By the way, on `bash` or `zsh` you could avoid explicitly naming a pipe by

```bash
securefs m --config <(gpg -d config.asc) priv pub
```

**Comment by antofthy** [comment]
Yes I know about that.  Bash itself passes the file descriptor file of the
command output to the command using the FD devices.
You can see it with...
   $ echo <(echo xyzzy)
   /dev/fd/63
   $ cat <(echo xyzzy)
   xyzzy
As long as the application only reads the argument ONCE, (as it is a
stream) all is fine.

Does not work well for non-bash applications (such as perl which it use to
manage encrypted filesystems, storing details (and the password & config)
of their encryption)

**Comment by WindProphet** [comment]
I try this on `bash` on macOS
```
securefs m --config <(gpg -d securefs.gpg) securefs /Volumes/securefs
```
but an error appear
```
[Error] [0x7fff8a08a380] [2018-04-16 01:54:26.811380000 UTC]    securefs::POSIXException: Illegal seek (lseek)
```

**Comment by netheril96** [comment]
Have you built `securefs` from source at master branch?

**Comment by schlomie** [comment]
This has been implemented and released in [0.8.2](https://github.com/netheril96/securefs/releases/tag/0.8.2)

Closing this...   Thank you @netheril96 - this works great!
